File: src-tauri/src/commands/github.rs
//! GitHub: the account, forks, pull requests and issues.

use crate::journal::{Action, Mode};
use crate::state::{blocking, in_repo, repo, watch_network, with_github, AppState, Res};
use crate::{git, github, network};
use std::path::PathBuf;
use tauri::ipc::Channel;
use tauri::{AppHandle, Manager, State};

/// https://github.com/owner/name when origin is on GitHub, for "Open on GitHub" links.
#[tauri::command]
pub async fn github_web_url(state: State<'_, AppState>) -> Res<Option<String>> {
    in_repo(&state, move |r| {
        Ok(git::remote_url(r, "origin")
            .and_then(|u| github::parse_remote(&u))
            .map(|g| format!("https://github.com/{}/{}", g.owner, g.name)))
    })
    .await
}

#[tauri::command]
pub async fn gh_account(app: AppHandle) -> Res<github::Account> {
    with_github(app, github::account).await
}

#[tauri::command]
pub async fn gh_original_remote(
    state: State<'_, AppState>,
    original: String,
    fetch: bool,
    op: String,
    progress: Channel<network::Progress>,
) -> Res<Option<String>> {
    let net = watch_network(&state, op, progress);
    in_repo(&state, move |r| {
        github::original_remote(r, &original, fetch, &net)
    })
    .await
}

#[tauri::command]
pub async fn gh_sync_fork(
    app: AppHandle,
    branch: String,
    op: String,
    progress: Channel<network::Progress>,
) -> Res<String> {
    let net = watch_network(&app.state::<AppState>(), op, progress);
    with_github(app, move |gh, r| github::sync_fork(gh, r, &branch, &net)).await
}

#[tauri::command]
pub async fn pull_draft(state: State<'_, AppState>, base: String) -> Res<git::PullDraft> {
    in_repo(&state, move |r| git::pull_draft(r, &base)).await
}

#[tauri::command]
pub async fn gh_remotes(state: State<'_, AppState>) -> Res<Vec<github::Remote>> {
    in_repo(&state, move |r| Ok(github::remotes(r))).await
}

#[tauri::command]
pub async fn gh_add_original_remote(
    app: AppHandle,
    op: String,
    progress: Channel<network::Progress>,
) -> Res<String> {
    let net = watch_network(&app.state::<AppState>(), op, progress);
    with_github(app, move |gh, r| github::add_original_remote(gh, r, &net)).await
}

#[tauri::command]
pub async fn gh_protected_branches(app: AppHandle) -> Res<Vec<String>> {
    with_github(app, github::protected_branches).await
}

#[tauri::command]
pub async fn pr_list(
    app: AppHandle,
    target: Option<String>,
    filter: String,
    pages: usize,
) -> Res<Vec<github::Pull>> {
    with_github(app, move |gh, r| {
        github::list(gh, r, target.as_deref(), &filter, pages)
    })
    .await
}

#[tauri::command]
pub async fn pr_review_comments(
    app: AppHandle,
    target: Option<String>,
    number: u64,
) -> Res<Vec<github::ReviewComment>> {
    with_github(app, move |gh, r| {
        github::review_comments(gh, r, target.as_deref(), number)
    })
    .await
}

/// A comment on a line of the PR's diff, or a reply in a thread (`reply_to`).
#[allow(clippy::too_many_arguments)]
#[tauri::command]
pub async fn pr_comment_line(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    commit: String,
    path: String,
    line: u64,
    side: String,
    reply_to: Option<u64>,
    body: String,
) -> Res<github::ReviewComment> {
    with_github(app, move |gh, r| {
        github::comment_line(
            gh,
            r,
            target.as_deref(),
            number,
            &commit,
            &path,
            line,
            &side,
            reply_to,
            &body,
        )
    })
    .await
}

/// The GitHub account's repositories, for cloning one; works with no repository open.
#[tauri::command]
pub async fn gh_own_repos(app: AppHandle) -> Res<Vec<github::OwnRepo>> {
    blocking(move || {
        let state = app.state::<AppState>();
        let cwd = repo(&state)
            .ok()
            .or_else(|| std::env::var_os("HOME").map(PathBuf::from))
            .unwrap_or_else(std::env::temp_dir);
        github::own_repos(&state.github, &cwd)
    })
    .await
}

/// CI's rollup for each commit GitHub has checks on (see github::ci_states).
#[tauri::command]
pub async fn ci_states(
    app: AppHandle,
    target: Option<String>,
    shas: Vec<String>,
) -> Res<std::collections::HashMap<String, String>> {
    with_github(app, move |gh, r| {
        github::ci_states(gh, r, target.as_deref(), &shas)
    })
    .await
}

#[tauri::command]
pub async fn pr_detail(
    app: AppHandle,
    target: Option<String>,
    number: u64,
) -> Res<github::PullDetail> {
    with_github(app, move |gh, r| {
        github::detail(gh, r, target.as_deref(), number)
    })
    .await
}

#[tauri::command]
pub async fn pr_attachments(
    app: AppHandle,
    target: Option<String>,
    number: u64,
) -> Res<std::collections::HashMap<String, String>> {
    with_github(app, move |gh, r| {
        github::attachments(gh, r, target.as_deref(), number)
    })
    .await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn pr_files(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    base_ref: String,
    base_sha: String,
    head_sha: String,
    op: String,
    progress: Channel<network::Progress>,
) -> Res<github::PullFiles> {
    let net = watch_network(&app.state::<AppState>(), op, progress);
    with_github(app, move |gh, r| {
        github::files(
            gh,
            r,
            target.as_deref(),
            number,
            &base_ref,
            &base_sha,
            &head_sha,
            &net,
        )
    })
    .await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn pr_create(
    app: AppHandle,
    target: Option<String>,
    title: String,
    body: String,
    head: String,
    base: String,
    draft: bool,
    maintainer_edits: bool,
) -> Res<github::Pull> {
    with_github(app, move |gh, r| {
        github::create(
            gh,
            r,
            target.as_deref(),
            &title,
            &body,
            &head,
            &base,
            draft,
            maintainer_edits,
        )
    })
    .await
}

#[tauri::command]
pub async fn pr_merge(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    method: String,
) -> Res<()> {
    with_github(app, move |gh, r| {
        github::merge(gh, r, target.as_deref(), number, &method)
    })
    .await
}

#[tauri::command]
pub async fn pr_set_open(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    open: bool,
) -> Res<github::Pull> {
    with_github(app, move |gh, r| {
        github::set_open(gh, r, target.as_deref(), number, open)
    })
    .await
}

#[tauri::command]
pub async fn pr_review(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    event: String,
    body: String,
) -> Res<()> {
    with_github(app, move |gh, r| {
        github::review(gh, r, target.as_deref(), number, &event, &body)
    })
    .await
}

#[tauri::command]
pub async fn issue_list(
    app: AppHandle,
    target: Option<String>,
    filter: String,
    labels: Vec<String>,
) -> Res<Vec<github::Issue>> {
    with_github(app, move |gh, r| {
        github::issues(gh, r, target.as_deref(), &filter, &labels)
    })
    .await
}

#[tauri::command]
pub async fn issue_counts(
    app: AppHandle,
    target: Option<String>,
    labels: Vec<String>,
) -> Res<github::IssueCounts> {
    with_github(app, move |gh, r| {
        github::issue_counts(gh, r, target.as_deref(), &labels)
    })
    .await
}

#[tauri::command]
pub async fn issue_labels(app: AppHandle, target: Option<String>) -> Res<Vec<github::Label>> {
    with_github(app, move |gh, r| {
        github::issue_labels(gh, r, target.as_deref())
    })
    .await
}

#[tauri::command]
pub async fn issue_detail(
    app: AppHandle,
    target: Option<String>,
    number: u64,
) -> Res<github::IssueDetail> {
    with_github(app, move |gh, r| {
        github::issue_detail(gh, r, target.as_deref(), number)
    })
    .await
}

#[tauri::command]
pub async fn issue_create(
    app: AppHandle,
    target: Option<String>,
    title: String,
    body: String,
) -> Res<github::Issue> {
    with_github(app, move |gh, r| {
        github::issue_create(gh, r, target.as_deref(), &title, &body)
    })
    .await
}

#[tauri::command]
pub async fn issue_edit(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    title: String,
    body: String,
) -> Res<github::Issue> {
    with_github(app, move |gh, r| {
        github::issue_edit(gh, r, target.as_deref(), number, &title, &body)
    })
    .await
}

#[tauri::command]
pub async fn issue_set_open(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    open: bool,
    reason: String,
) -> Res<github::Issue> {
    with_github(app, move |gh, r| {
        github::issue_set_open(gh, r, target.as_deref(), number, open, &reason)
    })
    .await
}

#[tauri::command]
pub async fn issue_set_labels(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    labels: Vec<String>,
) -> Res<Vec<github::Label>> {
    with_github(app, move |gh, r| {
        github::issue_set_labels(gh, r, target.as_deref(), number, &labels)
    })
    .await
}

#[tauri::command]
pub async fn issue_delete(app: AppHandle, target: Option<String>, number: u64) -> Res<()> {
    with_github(app, move |gh, r| {
        github::issue_delete(gh, r, target.as_deref(), number)
    })
    .await
}

#[tauri::command]
pub async fn issue_comment(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    body: String,
) -> Res<()> {
    with_github(app, move |gh, r| {
        github::issue_comment(gh, r, target.as_deref(), number, &body)
    })
    .await
}

#[tauri::command]
pub async fn pr_checkout(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    head_ref: String,
    same_repo: bool,
    op: String,
    progress: Channel<network::Progress>,
) -> Res<()> {
    let net = watch_network(&app.state::<AppState>(), op, progress);
    blocking(move || {
        let state = app.state::<AppState>();
        let r = repo(&state)?;
        let action = Action::new(format!("Check out PR #{number}"), Mode::Keep);
        state.journal.record(&r, action, |r| {
            let (remote, owner) = github::pr_source(&state.github, r, target.as_deref())?;
            github::checkout(
                r,
                &remote,
                owner.as_deref(),
                number,
                &head_ref,
                same_repo,
                &net,
            )
        })
    })
    .await
}

/// `pr_checkout` into a new worktree; returns its path. This worktree's HEAD doesn't move.
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn pr_checkout_worktree(
    app: AppHandle,
    target: Option<String>,
    number: u64,
    head_ref: String,
    same_repo: bool,
    dir: Option<String>,
    op: String,
    progress: Channel<network::Progress>,
) -> Res<String> {
    let net = watch_network(&app.state::<AppState>(), op, progress);
    blocking(move || {
        let state = app.state::<AppState>();
        let r = repo(&state)?;
        let (remote, owner) = github::pr_source(&state.github, &r, target.as_deref())?;
        github::checkout_worktree(
            &r,
            &remote,
            owner.as_deref(),
            number,
            &head_ref,
            same_repo,
            dir.as_deref(),
            &net,
        )
    })
    .await
}

File: src/lib/api/github.ts
import { invoke } from "@tauri-apps/api/core";
import { netOp, network } from "./network";
import type { FileChange, NetOp } from "./types";

/** What the signed-in account may do in one repository. */
export interface GitHubAccess {
  repo: { owner: string; name: string };
  defaultBranch: string | null;
  /** Write access: merge, and close or edit anyone's PRs and issues. */
  push: boolean;
  /** Triage: close and reopen anyone's PRs and issues, without write access. */
  triage: boolean;
  /** The only role GitHub lets delete issues. */
  admin: boolean;
  /** Issues are switched on (forks start with them off). */
  issues: boolean;
}

export interface GitHubAccount {
  login: string;
  /** "gh" = GitHub CLI, "git" = git's credential store */
  source: "gh" | "git";
  origin: GitHubAccess | null;
  /** The repository origin was forked from. */
  parent: GitHubAccess | null;
}

export const fullName = (r: GitHubAccess["repo"]) => `${r.owner}/${r.name}`;

/** owner/name from a PR's or issue's html url. */
export const repoOf = (url: string) => url.replace("https://github.com/", "").split("/").slice(0, 2).join("/");

/** The account's access in the repository a PR or issue (by url) belongs to. */
export function accessFor(account: GitHubAccount | null, url: string) {
  const repo = repoOf(url).toLowerCase();
  return [account?.origin, account?.parent].find((a) => a && fullName(a.repo).toLowerCase() === repo) ?? null;
}

/**
 * Where a request goes: null = origin, else the parent's owner/name. The backend accepts
 * nothing else. From a PR or issue, its own url says which.
 */
export type Target = string | null;

export interface Pull {
  number: number;
  title: string;
  state: "open" | "closed" | "merged";
  draft: boolean;
  author: string;
  headRef: string;
  headSha: string;
  headRepo: string | null;
  baseRef: string;
  baseSha: string;
  createdAt: string;
  updatedAt: string;
  url: string;
}

export interface PullCheck {
  name: string;
  state: string;
  url: string | null;
}

export interface PullComment {
  author: string;
  body: string;
  createdAt: string;
  review: string | null;
}

export interface PullDetail extends Pull {
  body: string;
  additions: number;
  deletions: number;
  changedFiles: number;
  commits: number;
  mergeable: boolean | null;
  mergeableState: string;
  checks: PullCheck[];
  /** Checks couldn't be read (a token without access to them, say); `checks` is then partial. */
  checksError: string | null;
  comments: PullComment[];
  /** Who closed it, if closed: an author may reopen only what they closed themselves. */
  closedBy: string | null;
  /** Who merged it, if merged: often not its author. */
  mergedBy: string | null;
}

export interface PullFiles {
  base: string;
  head: string;
  files: FileChange[];
}

export type MergeMethod = "merge" | "squash" | "rebase";
export type ReviewEvent = "APPROVE" | "REQUEST_CHANGES" | "COMMENT";

/** A page of the PR list (github.rs `PER_PAGE`). */
export const PR_PAGE = 100;

export type CiState = "success" | "failure" | "pending";

/** A comment on a line of a PR's diff; `line` null once the diff moved past it (outdated). */
export interface ReviewComment {
  id: number;
  replyTo: number | null;
  path: string;
  line: number | null;
  side: "LEFT" | "RIGHT";
  author: string;
  body: string;
  createdAt: string;
  url: string;
}

/** How many issues or pull requests are open and closed (a PR's closed counts the merged). */
export interface StateCounts {
  open: number;
  closed: number;
}

export const github = {
  account: () => invoke<GitHubAccount>("gh_account"),
  /** Origin's branches under branch protection (names without "origin/"). */
  protectedBranches: () => invoke<string[]>("gh_protected_branches"),
  /** The most recently updated `pages` × PR_PAGE. */
  list: (target: Target, filter: "open" | "closed" | "all", pages = 1) => invoke<Pull[]>("pr_list", { target, filter, pages }),
  /** All of them, not just the pages listed. */
  counts: (target: Target) => invoke<StateCounts>("pr_counts", { target }),
  detail: (target: Target, number: number) => invoke<PullDetail>("pr_detail", { target, number }),
  reviewComments: (target: Target, number: number) => invoke<ReviewComment[]>("pr_review_comments", { target, number }),
  /** On `line` of `path` at the PR's head `commit`, on `side`; with `replyTo`, an answer in that thread. */
  commentLine: (target: Target, number: number, commit: string, path: string, line: number, side: "LEFT" | "RIGHT", body: string, replyTo: number | null = null) =>
    invoke<ReviewComment>("pr_comment_line", { target, number, commit, path, line, side, replyTo, body }),
  /** The account's repositories (and those it works on), most recently updated first. */
  ownRepos: () => invoke<{ fullName: string; description: string; private: boolean; cloneUrl: string; updatedAt: string }[]>("gh_own_repos"),
  /** CI's rollup per commit, for those GitHub has checks on (up to 100 at once). */
  ciStates: (target: Target, shas: string[]) => invoke<Record<string, CiState>>("ci_states", { target, shas }),
  /** Signed image links for a private repo's attachments, by attachment id. */
  attachments: (target: Target, number: number) => invoke<Record<string, string>>("pr_attachments", { target, number }),
  /** Fetches the PR's commits first when they're missing. */
  files: (target: Target, p: Pick<Pull, "number" | "baseRef" | "baseSha" | "headSha">, op?: NetOp) =>
    network<PullFiles>("pr_files", { target, number: p.number, baseRef: p.baseRef, baseSha: p.baseSha, headSha: p.headSha }, op),
  /** Into the parent, `head` is a branch of origin's, and `maintainerEdits` lets its maintainers push to it. */
  create: (target: Target, title: string, body: string, head: string, base: string, draft: boolean, maintainerEdits = true) =>
    invoke<Pull>("pr_create", { target, title, body, head, base, draft, maintainerEdits }),
  merge: (target: Target, number: number, method: MergeMethod) => invoke<void>("pr_merge", { target, number, method }),
  setOpen: (target: Target, number: number, open: boolean) => invoke<Pull>("pr_set_open", { target, number, open }),
  review: (target: Target, number: number, event: ReviewEvent, body: string) => invoke<void>("pr_review", { target, number, event, body }),
  /** `sameRepo`: the PR's branch lives on origin, so it's checked out under its own name. */
  checkout: (target: Target, number: number, headRef: string, sameRepo: boolean, op?: NetOp) => network<void>("pr_checkout", { target, number, headRef, sameRepo }, op),
  /** The same branch, checked out in a new worktree in `dir` (default: beside the main one); returns its path. */
  checkoutWorktree: (target: Target, number: number, headRef: string, sameRepo: boolean, dir: string | null, op?: NetOp) =>
    network<string>("pr_checkout_worktree", { target, number, headRef, sameRepo, dir }, op),
  openUrl: (url: string) => invoke<void>("open_url", { url }),
  /** The remote for a fork's original, owner/name (fetched first if `fetch`); null when there is none. Works offline. */
  // Without `fetch` it's a local lookup: marked background, so it doesn't stop a background fetch.
  originalRemote: (original: string, fetch: boolean, op = netOp(undefined, !fetch)) => network<string | null>("gh_original_remote", { original, fetch }, op),
  /** This repo's remotes and the GitHub repositories (owner/name) behind them. Local only. */
  remotes: () => invoke<{ name: string; repo: string | null }[]>("gh_remotes"),
  /** GitHub's "Sync fork" for origin's `branch`, then a fetch of origin: "fast-forward", "merge" or "none". */
  syncFork: (branch: string, op?: NetOp) => network<"fast-forward" | "merge" | "none">("gh_sync_fork", { branch }, op),
  /** Adds the fork's original as a remote ("upstream") and fetches it; returns its name. */
  addOriginalRemote: (op?: NetOp) => network<string>("gh_add_original_remote", {}, op),
};
